addons/web/controllers/doc.py

```diff
@@ -0,0 +1,566 @@
+
+        # accumate the decorators such as @api.model
+        api = []
+        if getattr(method, '_api_model', False):
+            api.append('model')
+        if getattr(method, '_readonly', False):
+            api.append('readonly')
+
+        signature = parse_signature(method)
+        obj = {
+            **signature.as_dict(),
+            # signature: '',
+            # parameters: {},
+            # return: {}
+            # raise: {}
+            # doc: ''
+            'model': introducing_model._name or 'core',
+            'module': introducing_model._module or 'core',
+            # api: []
+        }
+        if api:
+            obj['api'] = api
+
+        return obj
+
+
+def get_sorted_installed_modules(env):
+    names = env['ir.module.module'].sudo().search([
+        ('state', '=', 'installed'),
+    ]).mapped('name')
+    graph = ModuleGraph(env.cr)
+    graph.extend(names)
+    return [p.name for p in graph]
+
+
+def is_public_method(model, name):
+    try:
+        return get_public_method(model, name)
+    except (AttributeError, AccessError):
+        return None
+
+
+CORE_FIELDS = ('id', 'display_name', 'create_uid', 'write_uid', 'create_date', 'write_date')
+def sort_key_field(sorted_module_list, model, field):  # noqa: E302
+    """
+    Key function to sort fields with the following order:
+
+    (1) core fields < (2) model fields < (3) custom ``x_`` fields
+
+    1. The core fields have a hardcoded order, like `id` is first.
+    2. The model fields are sorted first by *introducing module* (i.e.
+       base then web then website then ecommerce), and second by
+       alphabetical order.
+    3. The custom fields are sorted solely by alphabetical order.
+    """
+    if introducing_modules := model._fields[field['name']]._modules:
+        depth = sorted_module_list.index(introducing_modules[0])
+        return 2, depth, field['name']
+
+    if field['name'] in CORE_FIELDS:
+        return 1, CORE_FIELDS.index(field['name'])
+
+    assert field['name'].startswith('x_'), field['name']
+    return 3, field['name']
+
+
+def sort_key_method(sorted_module_list, model, method_name):
+    """
+    Key function to sort fields by the following criteria:
+
+    1) Depth of the module that introduced the method in the dependency grap.
+    2) Alphabetical order.
+    """
+    introducing_model = next(
+        parent_model
+        for parent_model
+        in reversed(type(model).mro())
+        if hasattr(parent_model, method_name)
+    )
+    if introducing_model._module:
+        depth = sorted_module_list.index(introducing_model._module)
+    else:
+        depth = -1
+    return depth, method_name
+
+
+def parse_signature(method) -> Signature:
+    isign = inspect.signature(method)
+
+    # strip self and cls from the signature
+    iter_params = iter(isign.parameters.items())
+    if next(iter_params, (None, None))[0] in ('self', 'cls'):
+        isign = isign.replace(parameters=(v for i, v in iter_params))
```

Comment:
It is actually quite hard to read, what about something like this?
```suggestion
    param_iter = iter(isign.parameters.values())
    for param in param_iter:
        if param.name in ('self', 'cls'):
            isign = isign.replace(parameters=param_iter)
        break
```

Or you could use `itertools.dropwhile` :)
```py
isign = isign.replace(parameters=itertools.dropwhile(lambda v: v.name in ('self', 'cls'), isign.parameters.values()))
```

Or even just iterator over `isign.parameters.values()`.